packages/calypso-e2e/src/lib/pages/signup/signup-pick-plan-page.ts

```diff
@@ -34,18 +37,31 @@ export class SignupPickPlanPage {
 		] );
 
 		let url: RegExp;
+		let actions: Array< Promise< any > > = [];
 		if ( name !== 'Free' ) {
 			// Non-free plans should redirect to the Checkout cart.
 			url = new RegExp( '.*checkout.*' );
 		} else {
 			url = new RegExp( '.*setup/site-setup.*' );
 		}
 
-		const actions = [
-			this.page.waitForResponse( /.*sites\/new\?.*/ ),
-			this.page.waitForURL( url, { timeout: 30 * 1000 } ),
-			this.plansPage.selectPlan( name ),
-		];
+		if ( name === 'Free' ) {
+			if ( this.selectedDomain?.includes( 'wordpress.com' ) ) {
+				/** Shows a modal */
+				await this.plansPage.selectPlan( name );
+				actions = [
+					this.page.waitForResponse( /.*sites\/new\?.*/ ),
+					this.page.waitForURL( url, { timeout: 30 * 1000 } ),
+					this.plansPage.selectModalUpsellPlan( name ),
+				];
+			}
+		} else if ( actions.length === 0 ) {
+			actions = [
+				this.page.waitForResponse( /.*sites\/new\?.*/ ),
+				this.page.waitForURL( url, { timeout: 30 * 1000 } ),
+				this.plansPage.selectPlan( name ),
+			];
+		}
```

Comment:
**question:blocking** - it looks like we're trying to handle a dialog box that can possibly appear, but is not guaranteed to. I think this is a fairly good attempt, but seems to repeat itself unnecessarily.

It isn't immediately clear from the code what should happen when the upsell happens. Is the intention to use the Free plan, or do we want to allow the user to click on the upsell plan?

I have a solution that removes much of this complexity, but assumes the user wants to revert to the Free plan. All existing flows that use `SignupPickPlanPage.selectPlan` don't sign up for a custom domain so there is no use for this if/else handling at the moment. Do you plan on writing a new E2E (which is a whole separate discussion to be had)?

Reply:
> Do you plan on writing a new E2E (which is a whole separate discussion to be had)?

Once we ship this and we are settled on the modal implementations (Maybe 1-2 months in after shipping and there are no red flags)
Then I will plan to write some E2E tests.  But for now the minimal changes to get the existing tests working should be enough.